lxd/storage/drivers/driver_cephfs.go

```diff
@@ -0,0 +1,977 @@
+		return err
+	}
+
+	if shared.PathExists(filepath.Join(mountPoint, fsPath)) {
+		// Delete the usual directories
+		for _, dir := range []string{"custom", "custom-snapshots"} {
+			if shared.PathExists(filepath.Join(mountPoint, fsPath, dir)) {
+				err = os.Remove(filepath.Join(mountPoint, fsPath, dir))
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Confirm that the path is now empty
+		ok, _ := shared.PathIsEmpty(filepath.Join(mountPoint, fsPath))
+		if !ok {
+			return fmt.Errorf("Only empty CEPHFS paths can be used as a LXD storage pool")
+		}
+
+		// Delete the path itself
+		if fsPath != "" && fsPath != "/" {
+			err = os.Remove(filepath.Join(mountPoint, fsPath))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// On delete, wipe everything in the directory.
+	err = wipeDirectory(GetPoolMountPath(d.name))
+	if err != nil {
+		return err
+	}
+
+	// Make sure the existing pool is unmounted
+	_, err = d.Unmount()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *cephfs) Mount() (bool, error) {
+	// Parse the namespace / path
+	fields := strings.SplitN(d.fsName, "/", 2)
+	fsName := fields[0]
+	fsPath := "/"
+	if len(fields) > 1 {
+		fsPath = fields[1]
+	}
+
+	// Get the credentials and host
+	monAddresses, secret, err := d.getConfig(d.config["cephfs.cluster_name"], d.config["cephfs.user_name"])
+	if err != nil {
+		return false, err
+	}
+
+	// Do the actual mount
+	connected := false
+	for _, monAddress := range monAddresses {
+		uri := fmt.Sprintf("%s:6789:/%s", monAddress, fsPath)
+		err = tryMount(uri, GetPoolMountPath(d.name), "ceph", 0, fmt.Sprintf("name=%v,secret=%v,mds_namespace=%v", d.config["cephfs.user_name"], secret, fsName))
+		if err != nil {
+			continue
+		}
+
+		connected = true
+		break
+	}
+
+	if !connected {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (d *cephfs) Unmount() (bool, error) {
+	return forceUnmount(GetPoolMountPath(d.name))
+}
+
+func (d *cephfs) GetResources() (*api.ResourcesStoragePool, error) {
+	// Use the generic VFS resources.
+	return vfsResources(GetPoolMountPath(d.name))
+}
+
+func (d *cephfs) ValidateVolume(vol Volume, removeUnknownKeys bool) error {
+	return d.validateVolume(vol, nil, removeUnknownKeys)
+}
+
+func (d *cephfs) CreateVolume(vol Volume, filler func(path string) error, op *operations.Operation) error {
+	if vol.volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	if vol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	volPath := vol.MountPath()
+
+	err := os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPath := true
+	defer func() {
+		if revertPath {
+			os.RemoveAll(volPath)
+		}
+	}()
+
+	if filler != nil {
+		err = filler(volPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	revertPath = false
+	return nil
+}
+
+func (d *cephfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool, op *operations.Operation) error {
```

Comment:
This appears to have the same ordering and wrapping issue as I fixed in `dir` yesterday.

The way to handle copy is:
 - Create volume (empty)
 - Copy content of first snapshot in
 - Make snapshot
 - Copy content of second snapshot in
 - Make snapashot
 - ...
 - Copy volume content

Snapshots cannot be written to so they need to be created from the current volume state.

Reply:
Done